File: engine/src/main/java/de/hanno/hpengine/util/ressources/FileMonitor.java
package de.hanno.hpengine.util.ressources;


import de.hanno.hpengine.engine.config.Config;
import de.hanno.hpengine.engine.config.SimpleConfig;
import org.apache.commons.io.monitor.FileAlterationMonitor;
import org.apache.commons.io.monitor.FileAlterationObserver;

public class FileMonitor {

//    TODO: Make this non-global
	private static FileMonitor instance;
	
	public FileAlterationMonitor monitor;
	public volatile boolean running = false;
	// TODO: Interface for hot reload deactivation...

    static {
		instance = new FileMonitor(500);
		try {
            if(SimpleConfig.isUseFileReloadXXX()) {
                instance.monitor.start();
                instance.running = true;
            }
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	public static FileMonitor getInstance() {
		return instance;
	}
	
	private FileMonitor(int interval) {
		monitor = new FileAlterationMonitor(interval);
	}
	
	public void add(FileAlterationObserver observer) {
        if(!SimpleConfig.isUseFileReloadXXX()) {
            return;
        }
		if (running) {
			try {
				observer.initialize();
				monitor.addObserver(observer);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}
	
	public void checkAndNotify() {
        if(!SimpleConfig.isUseFileReloadXXX()) {
            return;
        }
		if (!FileMonitor.getInstance().running) { return; }
		monitor.getObservers().forEach(o -> {o.checkAndNotify();});
	}

}

File: engine/src/main/java/de/hanno/hpengine/util/ressources/FileMonitor.kt
package de.hanno.hpengine.util.ressources

import org.apache.commons.io.FilenameUtils
import org.apache.commons.io.monitor.FileAlterationMonitor
import org.apache.commons.io.monitor.FileAlterationObserver
import java.io.File


object FileMonitor {
    private val directoryObservers = mutableMapOf<File, FileAlterationObserver>()
    val monitor: FileAlterationMonitor = FileAlterationMonitor(500).apply {
        start()
    }

    @JvmOverloads
    fun addOnFileChangeListener(file: File, overwriteShouldReload: (File) -> Boolean = { false }, action: (File) -> Unit) {
        val fileObserver = directoryObservers.getOrPut(file.parentFile) {
            FileAlterationObserver(file.parent).apply {
                initialize()
                monitor.addObserver(this)
            }
        }

        val listener: OnFileChangeListener = object: OnFileChangeListener() {
            fun shouldReload(changedFile: File): Boolean {
                val fileName = FilenameUtils.getBaseName(changedFile.absolutePath)
                return file.name.startsWith(fileName)
            }

            override fun onFileChangeAction(arg0: File) {
                if(overwriteShouldReload(arg0) || shouldReload(arg0)) {
                    action(arg0)
                }
            }
        }
        fileObserver.addListener(listener)
    }
}